Storage and Documentation v1.4 (#179)
CS2103-AY1819S2-W14-1/main: docs/DeveloperGuide.adoc

// tag::csvstorage[]
=== CSV Lessons storage feature
==== Current Implementation
The lesson CSV storage feature implements the following functions:

* Parsing lesson data into the a list of string arrays, which is a format opencsv accepts for saving to CSV.
* Parsing a list of string arrays back into lesson data, as retrieved from opencsv.

Actual file I/O is handled in the CsvUtil class, which uses the opencsv library to read/write CSV files.

==== CSV Lesson Format
.High-level sequence diagram for the `reloadLessons` command, an example of lesson loading
image::CsvLessonStorageSequenceDiagram.png[width="800"]

Lessons are parsed by CsvLessonListStorage, being converted between Lesson and List<String[]> formats. The List<String[]> format is the primary format handled by CsvUtil for reading and writing to CSV files.

For lessons, there are three main sections when saved to a CSV file.

===== Name

> Lesson names are equivalent to their file names.

Reviewer: Suggestion for clarity:

```suggestion
A lesson's name in BrainTrain is equivalent to its file name.
```

A lesson named "French" would be saved as French.csv, and vice-versa.

===== Header
The header is represented in the file as the first two lines. Each line represents a specific set of data, as outlined below.

The first line represents the type of the corresponding column. Each column can have three types:

. *Tested* +

> This marks the column as a core value. The first two instances of Tested in the file will be used in quizzes, and all remaining Tested values are treated as Not Tested.

Reviewer: ```suggestion
This marks the column as a core value. The first two instances of *Tested* in the file will be used in quizzes, and all remaining Tested values are treated as *Not Tested*.
```

[appendix]
== Use cases

[discrete]
=== Use case: Import a lesson

*MSS*

. User downloads file from a source (e.g Internet, other users).
. User moves the file to the lessons folder.

> . User enters `reloadLessons` command.

Reviewer: ```suggestion
. User enters `reload` command.
```
